lib/webidl2.js

```diff
@@ -522,63 +522,59 @@
       }
     }
 
-    function generic_type(typeName) {
-      const name = consume("FrozenArray", "Promise", "sequence", "record");
-      if (!name) {
-        return;
-      }
-      const ret = {
-        baseName: name.value,
-        escapedBaseName: name.value,
-        generic: { value: name.value, trivia: {} },
-        trivia: { base: name.trivia }
-      };
-      const open = consume("<") || error(`No opening bracket after ${name.type}`);
-      ret.generic.trivia.open = open.trivia;
-      switch (name.type) {
-        case "Promise": {
-          if (probe("[")) error("Promise type cannot have extended attribute");
-          const subtype = return_type(typeName) || error("Missing Promise subtype");
-          ret.idlType = [subtype];
-          break;
-        }
-        case "sequence":
-        case "FrozenArray": {
-          const subtype = type_with_extended_attributes(typeName) || error(`Missing ${name.type} subtype`);
-          ret.idlType = [subtype];
-          break;
+    class GenericType extends Type {
+      static parse(typeName) {
+        const base = consume("FrozenArray", "Promise", "sequence", "record");
+        if (!base) {
+          return;
         }
-        case "record": {
-          if (probe("[")) error("Record key cannot have extended attribute");
-          ret.idlType = [];
-          const keyType = consume(...stringTypes);
-          if (!keyType) error(`Record key must be a string type`);
-          const separator = untyped_consume(",") || error("Missing comma after record key type");
-          ret.idlType.push(Object.assign({ type: typeName }, EMPTY_IDLTYPE, {
-            baseName: keyType.value,
-            escapedBaseName: keyType.value,
-            idlType: keyType.value,
-            separator,
-            trivia: {
-              base: keyType.trivia
-            }
-          }));
-          const valueType = type_with_extended_attributes(typeName) || error("Error parsing generic type record");
-          ret.idlType.push(valueType);
-          break;
+        const ret = new GenericType({ tokens: { base } });
+        ret.tokens.open = optional_consume("<") || error(`No opening bracket after ${base.type}`);
+        switch (base.type) {
+          case "Promise": {
+            if (probe("[")) error("Promise type cannot have extended attribute");
+            const subtype = return_type(typeName) || error("Missing Promise subtype");
+            ret.subtype.push(subtype);
+            break;
+          }
+          case "sequence":
+          case "FrozenArray": {
+            const subtype = type_with_extended_attributes(typeName) || error(`Missing ${base.type} subtype`);
+            ret.subtype.push(subtype);
+            break;
+          }
+          case "record": {
+            if (probe("[")) error("Record key cannot have extended attribute");
+            const keyType = consume(...stringTypes) || error(`Record key must be a string type`);
```

Comment:
Only if not too annoying, might be nice: 
 
```suggestion
            const keyType = consume(...stringTypes) || error(`Record key must be one of: ${stringTypes.join(", ")}.`);
```